### ISSUE TITLE
AuthorizationHandler OR AuthorizeAttribute

### ISSUE THREAD

**Issue by jacobslusser** [aspnet/Security#1675]
This question is similar to #564... but I couldn't quite follow the discussion there. I currently have an `AuthorizationHandler` that checks for authorization based on convention, e.g. if role name matches the resource name. If that doesn't succeed, I would like it to then fallback to whatever was optionally specified in the `AuthorizeAttribute` of the controller action. 

Basically, I want an OR condition with my `AuthorizationHandler` and `AuthorizeAttribute`.

As it currently is, I can only get one or the other to work at the same time. Adding my custom `AuthorizationHandler` seems to short circuit the authorization logic and the `AuthorizeAttribute` doesn't seem to be getting checked.

**Comment by HaoK** [comment]
At the end of the day, everything boils down to a list of requirements that must be met.  The Authorize attribute by default just requires an authenticated user as its requirement.  

You can create a custom requirement and have a custom handler for your conventions, you can effectively get the behavior you describe by having your handler also succeed every other requirement when the convention is met (which no-ops the ones that would have been added from `[Authorize]`

**Comment by jacobslusser** [comment]
I think the "master requirement" you are describing is probably what I need to do. The only thing I need to do I think is know in my "master requirement" what all the other requirements are and enumerate them--including those that are specified in the `[Authorize(roles="")]`.

Any guidance?

**Comment by HaoK** [comment]
The handler is given the authorization context which has all of the requirements in it.  But you won't have any way to tell where they came from, and that's by design, you shouldn't care right?  If the 'master' convention succeeds, you want to authorize the request so just blindly succeed all pending requirements.

**Comment by HaoK** [comment]
So you actually don't need a master requirement at all, you just need to add a MasterHandler

**Comment by jacobslusser** [comment]
I don't think succeeding the other requirements is my problem (and that's a great idea by the way). It's when my requirement is not succeeded (but also not failed) where I want to include `Authorize` as fallback. Unless I'm misunderstanding you, you've suggested a way for step one to succeed step two, but not for step one to not succeed and step two to still succeed.

**Comment by HaoK** [comment]
Its subtle, but you can solve this problem without adding any special requirements.  If I understand your scenario properly, you want to allow all requests based on a convention, for the rest of the requests, you want them to go through the normal authorize attribute.

So starting for a normal app with working authorize, to add this new convention, you can just add a IAuthorizationHandler that checks to see if the convention is met, and if it is, succeed the authorization.  Otherwise do nothing which falls back to the existing authorize logic.  This is just how authorization works, all handlers are given a chance to authorize, your 'master' handler just allows you to override certain requests to return true.  It does not affect change the normal flow when the convention isn't met.

**Comment by HaoK** [comment]
Not quite, you want to register a custom IAuthorizationHandler that isn't tied to a specific requirement, you want to succeed the requirements for other handlers.

**Comment by HaoK** [comment]
Don't derive from `AuthorizationHandler<TReq>` in this case, you want to have more control here

**Comment by jacobslusser** [comment]
Ok, thanks @HaoK. Let me hack at it for a while and see what I can get going.

**Comment by jacobslusser** [comment]
Thanks @HaoK that was the help I needed. I think were I got confused was not knowing that an instance of `IAuthorizationHandler` could exist without a requirement and how exactly that would get called. When I was using `AuthorizationHandler<TReq>` I found that my code would not be called unless I registered the handler AND the requirement. I assumed that a requirement was necessary and that my `IAuthorizationHandler` would not run unless there was one.

Once I got a plain `IAuthorizationHandler` running it was trivial to man-handle the requirements in the way I wanted to.

**Comment by jacobslusser** [comment]
Actually, I still have a question on this...

When I enumerate the `PendingRequirements` (or even just `Requirements`) I expected to see a requirement for the `[Authorize(Roles = "")]` attribute in that list, e.g. the `Microsoft.AspNetCore.Authorization.Infrastructure.RolesAuthorizationRequirement` but I don't.

In order to get the behavior I want--which is to authorize via my custom requirement OR via the roles specified in the `Authorize` attribute--I need to be able to succeed that requirement if my custom requirement succeeds. However, it is not one of the ones I can enumerate.

What can I do?

**Comment by HaoK** [comment]
Can you check if your requirements are getting spread out into multiple authorize filters?  There's a bug that was fixed in 2.1 where the attributes weren't composing properly, which is fixed via a new CombineAuthorizePolicies setting in MVC.

**Comment by jacobslusser** [comment]
If I understand you correctly, yes, I am seeing multiple authorization filters. When I inspect the `AuthorizationFilterContext` list of `Filters`, there are two `AuthorizeFilters` in the list. One with a `DenyAnonymousAuthorizationRequirement` policy and one with the `RolesAuthorizationRequirement` policy. Be advised that I'm also using this startup code to require authenticated users (which I assume is the source of the `DenyAnonymousAuthorizationRequirement` filter):

```cs
var policy = new AuthorizationPolicyBuilder()
    .RequireAuthenticatedUser()
    .Build();

options.Filters.Add(new AuthorizeFilter(policy));
```
My Controller Action also has the `Authorize(Roles = "")` filter which I assume is the source of the `RolesAuthorizationRequirement`. However, only the `DenyAnonymousAuthorizationRequirement` is listed in the `PendingRequirements`, not the other one.
